WalletWasabi/WabiSabi/Models/Decomposition/DecompositionsOfASize.cs

```diff
@@ -0,0 +1,211 @@
+	// seconds on an i7) which is reasonable as a precomputation, but larger
+	// balances will need more storage as well as larger decompositions, so a
+	// more compact representation is desirable.
+	//
+	// Since the sizes are the same, they can be shifted at query time allowing
+	// for nominal values to be used instead of effective value which means that
+	// the combinations could be pre-computed once and saved to disk in a more
+	// compact representation for reuse between rounds even when considering
+	// large combinations and many random queries over a large effective value
+	// range (e.g. for coin selection).
+	internal record DecompositionsOfASize
+	{
+		public DecompositionsOfASize(IEnumerable<long> denominations, long Max, long Min)
+			: this(denominations.OrderByDescending(x => x).ToImmutableArray(), Max, Min) {}
+
+		private DecompositionsOfASize(ImmutableArray<long> denominations, long maximumEffectiveCost, long minimumEffectiveCost)
+		{
+			Denominations = denominations;
+
+			ByEffectiveCost = Denominations
+				.Select(x => new Decomposition(x))
+				.ToImmutableArray();
+
+			MaximumEffectiveCost = maximumEffectiveCost;
+
+			MinimumEffectiveCost = minimumEffectiveCost;
+		}
+
+		public ImmutableArray<Decomposition> ByEffectiveCost { get; private init; }
+
+		public ImmutableArray<long> Denominations { get; }
+
+		public int Count => ByEffectiveCost.Length;
+
+		public int Size => ByEffectiveCost[0].Outputs.Length;
+
+		public long MaximumEffectiveCost { get; private init; }
+
+		public long MinimumEffectiveCost { get; private init; }
+
+		// Returns the set of decompositions of the next size up by extending
+		// the decompositions of the current set with each of the base
+		// denominations.
+		//
+		// This is done by generating one set for each base value, which retains
+		// the order, and merging the resulting sets to preserve the global
+		// ordering.
+		internal DecompositionsOfASize Extend(long maximumEffectiveCost, long minimumEffectiveCost)
+			=> this with
+			{
+				MaximumEffectiveCost = Math.Min(MaximumEffectiveCost, maximumEffectiveCost),
+				MinimumEffectiveCost = Math.Max(MinimumEffectiveCost, minimumEffectiveCost),
+
+				ByEffectiveCost = Denominations
+				// FIXME
+				// AsParallel() has no apparent effect, parallelism only seems to make
+				// things unbearably slow if the inner `Extend` is also made
+				// AsParallel(), otherwise it seems to be stuck in a sequential
+				// execution mode. In addition to the below, also tried a
+				// partitioner.
+				// .AsParallel()
+				// .AsUnordered()
+				// .WithExecutionMode(ParallelExecutionMode.ForceParallelism)
+				// .WithMergeOptions(ParallelMergeOptions.NotBuffered)
+				// .WithDegreeOfParallelism(Denominations.Length)
+				.Select(x => Extend(x, maximumEffectiveCost, minimumEffectiveCost))
+				.Aggregate(ImmutableArray<Decomposition>.Empty as IEnumerable<Decomposition>, PossibleDecompositions.Merge)
```

Comment:
By forcing evaluation of certain enumerables in the process it can be made to evaluate some things in parallel at the expense of a very large memory footprint. In principle this does not require more than stack space logarithmic in the number of values, so the simplest way to parallelize this would be to spawn off a task for each level in a tree, that pulls items out of two channels/queues and pushes value into a single output.

For decomposition this doesn't seem to be worth the complexity, since these structures can be precomputed once and then reused between rounds all this would do is save a few seconds from that precomputation. However, for computing other users input combinations during output registration it might still be worth pursuing.